be/src/olap/rowset/segment_v2/binary_prefix_page.h

```diff
@@ -0,0 +1,175 @@
+        }
+        *reinterpret_cast<Slice*>(value) = Slice(_first_entry);
+        return Status::OK();
+    }
+
+    Status get_last_value(void* value) const override {
+        DCHECK(_finished);
+        if (_count == 0) {
+            return Status::NotFound("page is empty");
+        }
+        *reinterpret_cast<Slice*>(value) = Slice(_last_entry);
+        return Status::OK();
+    }
+
+   private:
+    PageBuilderOptions _options;
+    std::vector<uint32_t> _restart_points_offset;
+    faststring _first_entry;
+    faststring _last_entry;
+    size_t _count;
+    bool _finished;
+    faststring _buffer;
+};
+
+class BinaryPrefixPageDecoder : public PageDecoder {
+   public:
+    BinaryPrefixPageDecoder(Slice data, const PageDecoderOptions& options) :
+        _data(data), _parsed(false) {
+    }
+
+    Status init() override;
+
+    Status seek_to_position_in_page(size_t pos) override;
+
+    Status seek_at_or_after_value(const void* value, bool* exact_match) override;
+
+    Status next_batch(size_t* n, ColumnBlockView* dst) override;
+
+    size_t count() const override {
+        DCHECK(_parsed);
+        return _num_values;
+    }
+
+    size_t current_index() const override {
+        DCHECK(_parsed);
+        return _cur_pos;
+    }
+   private:
+    // decode shared and non-shared entry length from `ptr`.
+    // return ptr past the parsed value when success.
+    // return nullptr on failure
+    const uint8_t* _decode_value_lengths(const uint8_t* ptr, uint32_t* shared, uint32_t* non_shared);
+
+
+    // return start pointer of the restart point at index `restart_point_index`
+    const uint8_t* _get_restart_point(size_t restart_point_index) const {
+        return reinterpret_cast<const uint8_t*>(_data.get_data()) +
+               decode_fixed32_le(_restarts_ptr + restart_point_index * sizeof(uint32_t));
+    }
+
+    // read next value at `_cur_pos` and `_next_ptr` into `_current_value`.
+    // return OK and advance `_next_ptr` on success. `_cur_pos` is not modified.
+    // return NotFound when no more entry can be read.
+    // return other error status otherwise.
+    Status _read_next_value();
+
+    // seek to the first value at the given restart point
+    Status _seek_to_restart_point(size_t restart_point_index);
+
+    // copy `_current_value` into `output`.
+    Status _copy_current_to_output(MemPool* mem_pool, Slice* output);
+
+    Slice _data;
+    // byte offset in `_data` of the next value to read
+    // uint32_t _cur_offset;
+    bool _parsed;
+
+    // initialized in init()
+    size_t _num_values;
+    uint32_t _num_restarts;
+    // pointer to restart offsets array
+    const uint8_t* _restarts_ptr;
+    // ordinal of the first value to return in next_batch()
+    uint32_t _cur_pos;
+    // first value to return in next_batch()
+    faststring _current_value;
+    // pointer to the start of next value to read, advanced by `_read_next_value`
+    const uint8_t* _next_ptr;
```

Comment:
Better to make all members have default values.
```suggestion
    const uint8_t* _next_ptr = nullptr;
```

Reply:
done